File: creational/typescript/factory_method_structured/interfaces.ts
export interface Document {
  open(): void;
  close(): void;
  save(): void;
  getType(): string;
}

export abstract class Application {
  protected abstract createDocument(): Document;
  
  public newDocument(): void {
    const doc = this.createDocument();
    console.log(`Created: ${doc.getType()}`);
    doc.open();
    doc.save();
    doc.close();
  }
}

export interface PaymentProcessor {
  processPayment(amount: number): void;
  refundPayment(amount: number): void;
}

File: creational/dart/factory_method_structured/interfaces.dart
abstract class Document {
  void open();
  void close();
  void save();
  String get type;
}

abstract class DocumentApplication {
  Document createDocument();
  
  void newDocument() {
    var doc = createDocument();
    print('Created: ${doc.type}');
    doc.open();
    doc.save();
    doc.close();
  }
}

abstract class PaymentProcessor {
  void processPayment(int amount);
  void refundPayment(int amount);
}
